File: apps/app-frontend/src/helpers/skins.js
/**
 * Skin management service — all Tauri command wrappers.
 */
import { invoke } from '@tauri-apps/api/core'
import { fetch as tauriFetch } from '@tauri-apps/plugin-http'

export async function listSkins() {
  return await invoke('plugin:skins|skins_list')
}

export async function addSkin({ name, bytes, modelType, source, accountId, favorite }) {
  return await invoke('plugin:skins|skins_add', {
    name,
    bytes,
    modelType: modelType || 'classic',
    source: source || 'local',
    accountId: accountId || null,
    favorite: favorite || false,
  })
}

export async function deleteSkin(id) {
  return await invoke('plugin:skins|skins_delete', { id })
}

export async function updateSkin({ id, name, favorite, modelType, sortOrder, accountId }) {
  return await invoke('plugin:skins|skins_update', {
    id,
    name: name || null,
    favorite: favorite !== undefined ? favorite : null,
    modelType: modelType || null,
    sortOrder: sortOrder !== undefined ? sortOrder : null,
    accountId: accountId !== undefined ? accountId : null,
  })
}

export async function duplicateSkin(id) {
  return await invoke('plugin:skins|skins_duplicate', { id })
}

export async function reorderSkins(ids) {
  return await invoke('plugin:skins|skins_reorder', { ids })
}

export async function markSkinUsed(id) {
  return await invoke('plugin:skins|skins_mark_used', { id })
}

export async function validateSkin({ name, bytes }) {
  return await invoke('plugin:skins|skins_validate', { name: name || '', bytes })
}

/**
 * Convert a File (from <input> or drag/drop) to a byte array.
 */
export async function fileToBytes(file) {
  const buffer = await file.arrayBuffer()
  return new Uint8Array(buffer)
}

/**
 * Fetch a Minecraft player's skin by username or UUID using Mojang API.
 * Returns { name, uuid, skinUrl, skinBytes, modelType } or null.
 */
function fetchWithTimeout(url, ms = 10000) {
  const controller = new AbortController()
  const timer = setTimeout(() => controller.abort(), ms)
  return tauriFetch(url, { signal: controller.signal }).finally(() => clearTimeout(timer))
}

export async function lookupPlayerSkin(identifier) {
  try {
    // Determine if it's a UUID (with or without dashes) or username
    const isUuid = /^[0-9a-f]{32}$/i.test(identifier.replace(/-/g, ''))

    let profileUrl
    if (isUuid) {
      const uuid = identifier.replace(/-/g, '')
      profileUrl = `https://api.mojang.com/user/profiles/${uuid}`
    } else {
      profileUrl = `https://api.mojang.com/users/profiles/minecraft/${encodeURIComponent(identifier)}`
    }

    const profileResp = await fetchWithTimeout(profileUrl)
    if (!profileResp.ok) {
      if (profileResp.status === 204 || profileResp.status === 404) {
        throw new Error(`Player "${identifier}" not found`)
      }
      throw new Error(`Mojang API error (${profileResp.status})`)
    }

    const profile = await profileResp.json()
    if (!profile.id || !profile.name) {
      throw new Error('Invalid profile data')
    }

    // Fetch skin texture
    const sessionUrl = `https://sessionserver.mojang.com/session/minecraft/profile/${profile.id}`
    const sessionResp = await fetchWithTimeout(sessionUrl)
    if (!sessionResp.ok) {
      throw new Error('Could not fetch player profile data')
    }

    const session = await sessionResp.json()
    if (!session.properties) {
      return { name: profile.name, uuid: profile.id, skinUrl: null, modelType: 'classic' }
    }

    const textureProp = session.properties.find(p => p.name === 'textures')
    if (!textureProp) {
      return { name: profile.name, uuid: profile.id, skinUrl: null, modelType: 'classic' }
    }

    const textureData = JSON.parse(atob(textureProp.value))
    const skinEntry = textureData.textures?.SKIN
    if (!skinEntry) {
      return { name: profile.name, uuid: profile.id, skinUrl: null, modelType: 'classic' }
    }

    const modelType = skinEntry.metadata?.model === 'slim' ? 'slim' : 'classic'
    const skinUrl = skinEntry.url

    // Download the skin PNG
    const skinResp = await fetchWithTimeout(skinUrl)
    if (!skinResp.ok) {
      return { name: profile.name, uuid: profile.id, skinUrl, modelType }
    }

    const blob = await skinResp.blob()
    const buffer = await blob.arrayBuffer()
    const skinBytes = new Uint8Array(buffer)

    return { name: profile.name, uuid: profile.id, skinUrl, skinBytes, modelType }
  } catch (err) {
    throw err
  }
}

File: apps/app/src/api/skins.rs
use crate::api::Result;
use theseus::skins::{self, SkinModelType, SkinRecord};

pub fn init<R: tauri::Runtime>() -> tauri::plugin::TauriPlugin<R> {
    tauri::plugin::Builder::new("skins")
        .invoke_handler(tauri::generate_handler![
            skins_list,
            skins_add,
            skins_delete,
            skins_update,
            skins_duplicate,
            skins_reorder,
            skins_mark_used,
            skins_validate,
        ])
        .build()
}

#[tauri::command]
pub async fn skins_list() -> Result<Vec<SkinRecord>> {
    Ok(theseus::skins::list_skins().await?)
}

/// Add a new skin from raw PNG bytes
#[tauri::command]
pub async fn skins_add(
    name: String,
    bytes: Vec<u8>,
    model_type: Option<String>,
    source: Option<String>,
    account_id: Option<String>,
    favorite: Option<bool>,
) -> Result<SkinRecord> {
    Ok(theseus::skins::add_skin(
        &name,
        &bytes,
        model_type
            .as_deref()
            .map(SkinModelType::from_str)
            .unwrap_or(SkinModelType::Classic),
        source
            .as_deref()
            .map(theseus::skins::SkinSource::from_str)
            .unwrap_or(theseus::skins::SkinSource::Local),
        account_id.as_deref(),
        favorite.unwrap_or(false),
    )
    .await?)
}

#[tauri::command]
pub async fn skins_delete(id: String) -> Result<()> {
    Ok(theseus::skins::delete_skin(&id).await?)
}

#[tauri::command]
pub async fn skins_update(
    id: String,
    name: Option<String>,
    favorite: Option<bool>,
    model_type: Option<String>,
    sort_order: Option<i64>,
    account_id: Option<String>,
) -> Result<SkinRecord> {
    Ok(theseus::skins::update_skin(
        &id,
        name.as_deref(),
        favorite,
        model_type.as_deref().map(SkinModelType::from_str),
        sort_order,
        account_id.as_deref(),
    )
    .await?)
}

#[tauri::command]
pub async fn skins_duplicate(id: String) -> Result<SkinRecord> {
    Ok(theseus::skins::duplicate_skin(&id).await?)
}

#[tauri::command]
pub async fn skins_reorder(ids: Vec<String>) -> Result<()> {
    let refs: Vec<String> = ids;
    Ok(theseus::skins::reorder_skins(&refs).await?)
}

#[tauri::command]
pub async fn skins_mark_used(id: String) -> Result<()> {
    Ok(theseus::skins::mark_used(&id).await?)
}

/// Validates and adds a PNG skin file. Returns the created record.
#[tauri::command]
pub async fn skins_validate(
    name: String,
    bytes: Vec<u8>,
) -> Result<SkinRecord> {
    let (w, h) = theseus::skins::validate_skin_png(&bytes)?;
    let skin_name = if name.is_empty() {
        format!("Imported skin {w}x{h}")
    } else {
        name
    };
    let record = theseus::skins::add_skin(
        &skin_name,
        &bytes,
        SkinModelType::Classic,
        theseus::skins::SkinSource::Imported,
        None,
        false,
    )
    .await?;
    Ok(record)
}
